crmprtd/align.py

```diff
@@ -442,8 +442,12 @@ def align(sesh, row, diagnostic=False):
         )
         return None
 
-    # Create Obs object.
-    # Note: We are very specifically creating the Obs object here using the ids
-    # to avoid SQLAlchemy adding this object to the session as part of its
-    # cascading backref behaviour https://goo.gl/Lchhv6
-    return Obs(history_id=history.id, time=row.time, datum=datum, vars_id=var_id)
+    # To allow the use of the PostgreSQL dialect-specific `insert` function, we return a
+    # dict with the relevant key-value mapping. This may need to be converted to an Obs
+    # object, or not, depending on how it is used.
+    return {
+        "history_id": history.id,
+        "obs_time": row.time,
+        "datum": datum,
+        "vars_id": var_id,
+    }
```

Comment:
I don't like this change of interface between `align` and `insert` for a couple reasons.

1. To me, it feels like a leaky abstraction. We want a clean interface between the `align` and `insert` phases, but then the fact that we still do more conversions on the data after they go to `insert` (or not) feels... messy to me. Though I'm having trouble articulating why its necessarily problematic. I guess to me the "promise" of `align` has always been, "I'm passing you something that theoretically should be able to go into the database", and a `dict` feels much, much less guaranteed.
2. It's loosely specified. "a dict" can be or contain anything. It's kind of like saying: "give me your data in JSON and everything will work!" (but it doesn't). This can be problematic both for the programmer ("what does the dict need to contain?") and for the Python interpreter which *could* adaptively optimize for a more tightly constrained data structure, but cannot for a loosely and openly defined `dict`. For most things the performance may not matter, but for the central data being passed between pieces of the program, it may.

If we think that we want to not use the ORM's `Obs` class, than maybe we should at least create a [`NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple), or some other more specified and immutable object.

Reply:
I thought about this. I considered both your objection/solution and the inverse of what's here (i.e., return Obs, and add a conversion from Obs to a dict for the bulk insertion with pg_insert). I prefer the second; it's cleaner and more in the spirit of SQLAlchemy. All I can say is I was working quickly ...